### ISSUE TITLE
Linux support for ACAP Lib

### ISSUE THREAD

**Issue by sumukhbansal** [gaolinorange/Automatic-Unpaired-Shape-Deformation-Transfer#9]
Dear Authors,
Congratulations on the beautiful work.
Is there a Linux version of the ACAP code available? 
Regards,

**Comment by MaxMayya** [comment]
Looking for that too. @sumukhbansal Did you find a solution for it?

**Comment by sumukhbansal** [comment]
@MaxMayya No. I also tried to implement the paper, but it's hard to reproduce.

**Comment by MaxMayya** [comment]
@sumukhbansal thanks for the quick reply. Hope authors come with an update soon!

**Comment by tommaoer** [comment]
@MaxMayya @sumukhbansal Thanks for your interest. We have a Win version(x64) of the ACAP [code](https://github.com/gaolinorange/Automatic-Unpaired-Shape-Deformation-Transfer/tree/67d86c91e0771c575dfbf24da476c31584099f1e/ACAP). We will release a Linux version in a few days. Thanks for your patience.

**Comment by MaxMayya** [comment]
@tommaoer great news! looking forward to that release.
Thanks a lot for your responsiveness.

**Comment by tommaoer** [comment]
@MaxMayya The code actually is the implementation of ACAP, which is a part of our sparse deformation [algorithm](https://arxiv.org/abs/1709.01250). The sparse deformation is based on [ARAP](https://igl.ethz.ch/projects/ARAP/arap_web.pdf). So we call the "ACAP" as "new_ARAP" in the early phase of the project. But I think it's not a problem, and we will fix it for easily reading.

**Comment by MaxMayya** [comment]
@tommaoer Thanks a lot for the clarification. This makes more sense.

**Comment by tommaoer** [comment]
@sumukhbansal @MaxMayya hi, it's a news that we have shared the Linux version of ACAP code! If you have further problems, please feel free to contact us.

**Comment by MaxMayya** [comment]
@tommaoer Many thanks for sharing the linux version! Looks neat! 
I've noticed that you use Gurobi which needs a license. Accepting sacrificing performance, can one run the code without Gurobi?
Appreciate the effort

**Comment by tommaoer** [comment]
@tommaoer Yes, thanks to the better performance, we solve the mixed-integer optimization by Gurobi. You can use other libraries to achieve optimization. For the academic license, it's free (3 months) for academic organizations.

For our released code, you can calculate the ACAP feature without the optimization, which is the original deformation gradient.

**Comment by MaxMayya** [comment]
I could ask for a free trial as I'm not attached to any academic organisation but let's say I want to run the code without optimisation. Typically using `./main path-to-folder -opt 0` would use the original deformation gradient right? 

I don't know if this is a related issue but after building everything (without the license `grbgetkey` call), calling ./main ../../test -opt 0 (or 1) ends up with the following error:

> ./main: error while loading shared libraries: libgurobi80.so: cannot open shared object file: No such file or directory 

I checked that the **libgurobi80.so** is physically in the right place. So I'm not sure if this is a random issue or related to the license key. Any hints?

**Comment by tommaoer** [comment]
@MaxMayya If you only use the original deformation gradient, the mixed-integer optimization is unavailable. Hence, I think you can remove the GUROBI and Comiso library and its relative code. Then, re-compile the library to run it. You can try this. If you encounter some problems, you can let me know.

**Comment by tommaoer** [comment]
hi, thanks for your remind. This is an accident I uploaded two wrong files. We have uploaded the right files on the same path. Please check it. I will check if this error is caused by my code.

**Comment by MaxMayya** [comment]
Thanks for updating the files so quickly! Now optimized rotation angles `theta` is coherent with their initial values.
I tested the groundtruth `theta` for the vertex **1153** on **2.obj** (last vertex of the cylinder) and found `18 deg`. However, when I visualize **2.obj** I can intuitively tell that vertex **1153** is rotated by `(360-18) deg`. Can you confirm?
BTW, can you kindly provide a tool (or a way) to inverse the ACAP feature vector back to the target mesh?
Thank you again

**Comment by tommaoer** [comment]
Sorry for the delayed reply. I check the optimized the 'logr' for the last vertex 1153.
```
         0       5.9691 0
Log(R) = -5.9691 0      0
         0       0      0
```
The optimized the direct of axis is consistent. so the rotation angle is 5.9691. In other words, the rotation angle may `2pi - 5.9691' when the direct of axis change. BTW, the provided 'LOGRNEW_GT' and 'S_GT' is without optimization, and also we have uploaded the 'GT' files with optimization. If you need it, please check it in same folder.

We also provide a complied script as follow line for reconstruct the ACAP feature back to the coordinates.
https://github.com/gaolinorange/Automatic-Unpaired-Shape-Deformation-Transfer/blob/39eec6c18a744f9bf061beaaf600dd855059b14c/matlabscript/recon_from_random.m#L94

If you have some question on the above script, please feel free to contact us.

**Comment by MaxMayya** [comment]
I appreciate taking time to check this in details.
I believe that you accidentally inverted file names in your last update (because I get identical results as _LOGRNEW_withopt_GT_ and _S_withopt_GT_ when I call the `main` with `-opt 0`).
Thank you for the pointer on the script. Unfortunately, I don't have Matlab :(, so I cannot test the inverse of ACAP! BTW, does the inverse of ACAP depend on wether the feature vector was computed with or without optimisation?

**Comment by tommaoer** [comment]
Thank you for your reminder. The inverse of ACAP does not depend on the optimization, it is suitable for the feature vector that is computed with or without optimization. 
BTY, if you have a windows system, we can provide a complied exe for the ACAP feature reconstruction for convenience. We will provide the source code for the reconstruction of the ACAP feature because I am working on the other research. :)

**Comment by MaxMayya** [comment]
Unfortunately, I don't have a Windows system (otherwise I wouldn't bother you with all the linux stuff ^^). A code source for the reconstruction from ACAP is more than welcome!
Thank you a lot for your availability

**Comment by MaxMayya** [comment]
@tommaoer back again with some news:
I have sloved, for the test vector, the axis problem using coin-or _Couenne_ solver and the angle problem using _Bonmin_. Without adding time limits or specifying any gaps, solvers had reached optimality after ~4.5min (30 sec for axis problem and 4 min for the angle problem). Can you share computation time values using the Gurobi optimizer?
_Side note_: I think you accidentally inverted the names of the opt/noOpt test vectors.

**Comment by mingliangfu** [comment]
> @MaxMayya The code actually is the implementation of ACAP, which is a part of our sparse deformation [algorithm](https://arxiv.org/abs/1709.01250). The sparse deformation is based on [ARAP](https://igl.ethz.ch/projects/ARAP/arap_web.pdf). So we call the "ACAP" as "new_ARAP" in the early phase of the project. But I think it's not a problem, and we will fix it for easily reading.

Hi, @tommaoer, could your share the rest part of sparse deformation algorithm?
